File: packages/shared/src/worldspec.ts
import { z } from 'zod';

const compositionSchema = z.record(z.string(), z.number().nonnegative());

export const worldSpecSchema = z.object({
  schemaVersion: z.literal(1),
  seed: z.string().min(1),
  star: z.object({
    spectralClass: z.string(),
    massSolar: z.number().positive(),
    luminositySolar: z.number().positive(),
    ageGyr: z.number().nonnegative(),
    binaryCompanion: z
      .object({
        spectralClass: z.string(),
        separationAu: z.number().positive(),
        eccentricity: z.number().min(0).max(1)
      })
      .optional()
  }),
  orbit: z.object({
    semiMajorAxisAu: z.number().positive(),
    eccentricity: z.number().min(0).max(1),
    inclinationDeg: z.number(),
    axialTiltDeg: z.number().min(0).max(180),
    rotationPeriodHours: z.number().positive(),
    tidalLock: z.enum(['none', 'primary', 'resonant'])
  }),
  body: z.object({
    massEarth: z.number().positive(),
    radiusEarth: z.number().positive(),
    bulkDensityKgM3: z.number().positive(),
    tectonics: z.enum(['plate', 'stagnant-lid', 'episodic', 'cryovolcanic', 'inert']),
    composition: z.object({ core: z.string(), mantle: z.string(), crust: z.string() }),
    magneticFieldMicrotesla: z.number().nonnegative()
  }),
  volatiles: z.record(z.string(), z.number().nonnegative()),
  surface: z.object({
    elevationHint: z.enum(['smooth', 'mixed', 'rugged']),
    dominantMaterials: z.array(z.string()).min(1),
    impactHistory: z.enum(['quiet', 'moderate', 'heavy'])
  }),
  atmosphere: z.discriminatedUnion('kind', [
    z.object({ kind: z.literal('none') }),
    z.object({
      kind: z.literal('present'),
      pressureBar: z.number().positive(),
      composition: compositionSchema
    })
  ]),
  hydrosphere: z.discriminatedUnion('kind', [
    z.object({ kind: z.literal('none') }),
    z.object({ kind: z.literal('water-ocean'), coverage: z.number().min(0).max(1), salinityPsu: z.number().nonnegative() }),
    z.object({ kind: z.literal('methane-lakes'), coverage: z.number().min(0).max(1) }),
    z.object({ kind: z.literal('brine'), coverage: z.number().min(0).max(1), salinityPsu: z.number().nonnegative() }),
    z.object({ kind: z.literal('lava'), coverage: z.number().min(0).max(1) })
  ]),
  life: z.discriminatedUnion('kind', [
    z.object({ kind: z.literal('none') }),
    z.object({ kind: z.literal('microbial') }),
    z.object({ kind: z.literal('complex') }),
    z.object({ kind: z.literal('sapient'), techLevel: z.enum(['preindustrial', 'industrial', 'digital', 'spacefaring']), culturalSeed: z.string() })
  ]),
  moons: z.array(
    z.object({
      name: z.string(),
      massEarth: z.number().positive(),
      radiusEarth: z.number().positive(),
      orbitSemiMajorAxisKm: z.number().positive(),
      composition: z.enum(['rocky', 'icy', 'mixed'])
    })
  ),
  rings: z
    .object({
      innerRadiusKm: z.number().positive(),
      outerRadiusKm: z.number().positive(),
      composition: z.enum(['ice', 'dust', 'mixed'])
    })
    .optional()
});

export type WorldSpec = z.infer<typeof worldSpecSchema>;
export type Archetype =
  | 'earth-analog'
  | 'mars-like'
  | 'venus-like'
  | 'titan-like'
  | 'europa-like'
  | 'airless-rockball'
  | 'tide-locked-m-dwarf-desert';

File: crates/sim-core/src/worldspec.rs
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct WorldSpec {
    pub schemaVersion: u8,
    pub seed: String,
    pub star: Star,
    pub orbit: Orbit,
    pub body: Body,
    pub volatiles: BTreeMap<String, f64>,
    pub surface: Surface,
    pub atmosphere: Atmosphere,
    pub hydrosphere: Hydrosphere,
    pub life: Life,
    pub moons: Vec<MoonSpec>,
    pub rings: Option<Rings>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Star { pub spectralClass: String, pub massSolar: f64, pub luminositySolar: f64, pub ageGyr: f64, pub binaryCompanion: Option<BinaryCompanion> }
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct BinaryCompanion { pub spectralClass: String, pub separationAu: f64, pub eccentricity: f64 }
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Orbit { pub semiMajorAxisAu: f64, pub eccentricity: f64, pub inclinationDeg: f64, pub axialTiltDeg: f64, pub rotationPeriodHours: f64, pub tidalLock: String }
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Body { pub massEarth: f64, pub radiusEarth: f64, pub bulkDensityKgM3: f64, pub tectonics: String, pub composition: Composition, pub magneticFieldMicrotesla: f64 }
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Composition { pub core: String, pub mantle: String, pub crust: String }
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Surface { pub elevationHint: String, pub dominantMaterials: Vec<String>, pub impactHistory: String }
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "kind")]
pub enum Atmosphere { #[serde(rename = "none")] None, #[serde(rename = "present")] Present { pressureBar: f64, composition: BTreeMap<String, f64> } }
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "kind")]
pub enum Hydrosphere {
    #[serde(rename = "none")] None,
    #[serde(rename = "water-ocean")] WaterOcean { coverage: f64, salinityPsu: f64 },
    #[serde(rename = "methane-lakes")] MethaneLakes { coverage: f64 },
    #[serde(rename = "brine")] Brine { coverage: f64, salinityPsu: f64 },
    #[serde(rename = "lava")] Lava { coverage: f64 },
}
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "kind")]
pub enum Life { #[serde(rename = "none")] None, #[serde(rename = "microbial")] Microbial, #[serde(rename = "complex")] Complex, #[serde(rename = "sapient")] Sapient { techLevel: String, culturalSeed: String } }
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct MoonSpec { pub name: String, pub massEarth: f64, pub radiusEarth: f64, pub orbitSemiMajorAxisKm: f64, pub composition: String }
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Rings { pub innerRadiusKm: f64, pub outerRadiusKm: f64, pub composition: String }
